Updates on full reread (#25)
wheelnext/peps: peps/pep-0815.rst

The Python wheel packaging format uses
:doc:`packaging:specifications/platform-compatibility-tags` to specify
a given wheel’s supported environments. These tags are unable to express

> the features of modern hardware configurations, such as the availability of

Reviewer: ```suggestion
features of modern hardware configurations, such as the availability of
```

We're referring only to specific features that modern hardware has, not all of them.

**GPU Accelerated Frameworks:** A wheel filename like
``torch-2.9.0-cp313-cp313-manylinux_2_28_x86_64.whl`` provides no

> indication whether it provides the support for NVIDIA CUDA, AMD ROCm,

Reviewer: ```suggestion
indication whether it supports NVIDIA CUDA, AMD ROCm,
```

Separate package indexes as variants
''''''''''''''''''''''''''''''''''''

> Projects like `PyTorch <https://pytorch.org/get-started/locally/>`__

Reviewer: ```suggestion
Projects such as `PyTorch <https://pytorch.org/get-started/locally/>`__
```

    pip install torch --index-url https://download.pytorch.org/whl/cu129

Tools need to implement special handling for the way PyTorch uses local

> version segments. PyTorch breaks the pattern that packages are usually

Reviewer: This was intentionally phrased around torch, cause it's such as dominant problem.

Author: Sure, but the sentence was really hard to understand because of "packages are usually installed with through these requirements".

Reviewer 2: The paragraph is still about PyTorch, and I agree that the original phrasing was harder to understand. So I think this is an improvement and can be merged.

are usually installed with. Problems with installing PyTorch
are a very common point of user confusion. To quantify this, on

> 2025-12-05 the 552 out of 8136, or 6.8%, of issues on the `uv's issue

Reviewer: ```suggestion
2025-12-05 552 out of 8136, or 6.8%, of issues on the `uv's issue
```

Author: Sure, but we need a comma there.

> tracker <https://github.com/astral-sh/uv/issues/>`__ contain the term

Reviewer: ```suggestion
tracker <https://github.com/astral-sh/uv/issues/>`__ contained the term
```

Package names as variants
'''''''''''''''''''''''''

Packages such as `XGBoost
<https://xgboost.readthedocs.io/en/stable/install.html>`__ use different
package names to approximate variants:

.. code:: bash

    pip install xgboost      # NVIDIA GPU variant
    pip install xgboost-cpu  # CPU-only variant

However, in that case maintainers of other software cannot cleanly
express a dependency on any of the available variants. They need to
either depend on a specific variant, provide multiple alternative
dependency sets using extras, or even publish their own software using

> multiple package names matching upstream variants.

Reviewer: ```suggestion
Maintainers of other software cannot express that they depend on
exactly one of the available variants being selected. They need to
either depend on a specific variant, provide multiple alternative
dependency sets using extras, or even publish their own software using
multiple package names matching upstream variants.
```

Author: Maybe without "exactly", or "either" instead. Otherwise, it sounds like they want to depend on `xgboost` or `xgboost-cpu`, which is what they can do.

Commonly, these packages install overlapping files. Since Python
packaging does not support expressing that two packages are mutually
exclusive, installers can install both of them to the same environment,
with the package installed second overwriting files from the one

> installed first. This could lead to unpredictable behavior, including

Reviewer: ```suggestion
installed first. This could leads to runtime errors, and
```

We've seen those on the uv issue tracker, it's a real problem.

> installed first. This could leads to runtime errors, and

Reviewer 2: ```suggestion
installed first. This could lead to runtime errors, and
```

Author: Actually, I think @konstin meant "this leads".

``numpy-cuda`` package that users will be lead to believe it’s a CUDA
variant of NumPy.

> For diverse reasons, `CuPy <https://cupy.dev/>`__ had to build

Reviewer: I'd drop the "For diverse reasons", if we don't list the reasons, it doesn't help to say that such reasons exist, that's already implied by the "had to".

compatibility. For example, a package that features optional GPU support
could publish the following wheels:

- One or more wheel variants built for specific hardware, that will be

>   installed on systems with an installer supporting variants and

Reviewer: I wouldn't add that, our assumption is that installers will support variants. If users can't rely that a mainstream tool that they are using is supporting variants, we IMHO failed a goal of this PEP, cause then we can't rely on variants in the ecosystem.

Author: It's for the transition period.

   provide.

2. Variants that can always be installed, and therefore only provide

>    a choice between different installation options. For example, software built

Reviewer: I think the `a` is correct here

Author: It sounded weird to me. I mean, it's not just one choice, it's choice in general.

Reviewer 2: "a choice" is better, I'd revert this

Author: Done.

CUDA variants with different minimal CUDA runtime versions and supported
GPUs, two ROCm variants and a Linux XPU variant.

> This setup could be improved using three GPU/XPU plugins that query the

Reviewer 2: Applying a suggestion made on the staging PR to drop the number completely:

```suggestion
This setup could be improved using GPU/XPU plugins that query the
```
